### ISSUE TITLE
SplitChunksPlugin and stats.toJson cause slow incremental watching compile

### ISSUE THREAD

**Issue by hedgepigdaniel** [webpack/webpack#9718]
# Bug report

**What is the current behavior?**

In a project with 5500 modules, incremental compile time after changing a single module is 6-7s

**What is the expected behavior?**

Faster is better

**Other relevant information:**
webpack version:  4.39.2
Node.js version: 10.16.3
Operating System: Linux

![image](https://user-images.githubusercontent.com/1217649/65289575-96999f80-db8e-11e9-9c59-b47439c09b88.png)

![image](https://user-images.githubusercontent.com/1217649/65289598-b761f500-db8e-11e9-9c4d-481b2d43ecc2.png)

![image](https://user-images.githubusercontent.com/1217649/65289994-75d24980-db90-11e9-83c6-45fec8b1a964.png)

See attached profiles.

In particular, I wonder if there are some simple changes that would reap large gains:
- getModulesSize and addModuleToChunksInfoMap in SplitChunksPlugin take 2068ms and 845ms respectively, mostly by indirectly calling size() in NormalModule. Can this be cached? I only changed a single source file, so I can't see why it needs to go back and remeasure the length of the source or whatever for many modules.
- There seems to be alot of sorting going on in stats.toJson - is this really necessary?

**Comment by sokra** [comment]
> * getModulesSize and addModuleToChunksInfoMap in SplitChunksPlugin take 2068ms and 845ms respectively, mostly by indirectly calling size() in NormalModule. Can this be cached? I only changed a single source file, so I can't see why it needs to go back and remeasure the length of the source or whatever for many modules.

yes, PR merged.

> * There seems to be alot of sorting going on in stats.toJson - is this really necessary?

yes, json should be deterministic.

The strategy here is to avoid requesting all stats data and only choose what you need. See argument of `toJson`, resp. https://github.com/FormidableLabs/webpack-stats-plugin#custom-stats-configuration

You can also opt out of most sorting with i. e. `stats.modulesSort` option.

**Comment by hedgepigdaniel** [comment]
Ok thanks, between those things I've got it down to about 3-3.5s which is much better. Turns out stats.toJson was getting called twice, and in both cases I didn't need the module/chunk information at all (just assets, which is fast). One of those I've [fixed](https://github.com/hedgepigdaniel/webpack-hmr-client/pull/3/commits/b68a903ed0d11b6188db117321329085e1be900e) in [webpack-hmr-client](https://github.com/hedgepigdaniel/webpack-hmr-client) (my fork of webpack-hot-client).

I don't see any other easy/obvious ways to improve it. In my case most of the time is still spent in SplitChunksPlugin. I suspect it would be possible to reuse most of the data structure it builds from the compilation and avoid traversing all modules and rebuilding it each time, but that's not something I can realistically work on.
